kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionsIterator.java

```diff
@@ -179,19 +181,85 @@ private CloseableIterator<ActionWrapper> getNextActionsIter() {
-
+                List<FileStatus> checkpointFiles =
+                        retrieveRemainingCheckpointFiles(fileVersion);
+
+                // If the checkpoint file is a UUID-named checkpoint, we may need to read the JSON
+                // actions separately. Otherwise, we can add the checkpoint manifest to the list of
+                // checkpoint files to be read in parallel.
+                CloseableIterator<ColumnarBatch> checkpointJsonActions;
+                if (checkpointInstance.format.equals(CheckpointInstance.CheckpointFormat.V2) &&
+                    nextFilePath.getName().endsWith(".json")) {
+                    checkpointJsonActions = tableClient.getJsonHandler().readJsonFiles(
+                            toCloseableIterator(
+                                    Collections.singletonList(nextFile).iterator()),
+                            readSchema,
+                            Optional.empty());
+                } else {
+                    checkpointFiles.add(nextFile);
+                    checkpointJsonActions = toCloseableIterator(
+                            new ArrayList<ColumnarBatch>().iterator());
+                }
+
+                // Read any potential sidecars for a checkpoint format supporting sidecars and if
+                // the current read is for Add/Remove Files (otherwise, no reason to read sidecars).
+                if (checkpointInstance.format.usesSidecars() &&
+                        LogReplay.isAddRemoveReadSchema(readSchema)) {
+                    // If the checkpoint format may contain sidecars, read the sidecar actions from
+                    // the checkpoint.
+                    final CloseableIterator<ColumnarBatch> manifestIter;
+                    if (nextFilePath.getName().endsWith(".parquet")) {
+                        manifestIter = tableClient.getParquetHandler().readParquetFiles(
+                                toCloseableIterator(
+                                        Collections.singletonList(nextFile).iterator()),
+                                LogReplay.getSidecarFileSchema(),
+                                Optional.empty());
+                    } else if (nextFilePath.getName().endsWith(".json")) {
+                        manifestIter = tableClient.getJsonHandler().readJsonFiles(
+                                toCloseableIterator(
+                                        Collections.singletonList(nextFile).iterator()),
+                                LogReplay.getSidecarFileSchema(),
+                                Optional.empty());
+                    } else {
+                        throw new IOException(String.format("Unrecognized file format for file: %s",
+                                nextFilePath.getName()));
+                    }
+
+                    // Read SidecarFile actions and add the corresponding files to the list of
+                    // checkpoint files to be read.
```

Comment:
what about the alternative of:

Create closeable iterator of asked actions + side car column from uuid-based checkpoint. Add a wrapper around it to collect the `sidecar` files as it is being read. On the next getNextActionsIter call issue one read request for all `sidecar` files. This avoids the reading of the uuid file twice.

Reply:
Since the actions are returned in `ColumnarBatch` of actions, how would we handle removing the sidecar files from the iterator? Would we have to make a new `ColumnarBatch` with the sidecars filtered out?

Reply:
you can use the API [here](https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/data/ColumnarBatch.java#L80) to delete a specific column from the existing batch.

Reply:
Got it - the one downside then would be that there would be some rows that are all `null` then, correct?